File: SRM 682/SmilesTheFriendshipUnicorn.cpp
#include <cstdio>
#include <cmath>
#include <cstring>
#include <ctime>
#include <iostream>
#include <algorithm>
#include <set>
#include <vector>
#include <sstream>
#include <typeinfo>
#include <fstream>

using namespace std;

class SmilesTheFriendshipUnicorn {
    public:
    vector<int> adj[10005];
    bool vis[10005];
    bool dfs(int u, int d) {
        if (d == 5) return true;
        vis[u] = true;
        for (int i = 0; i < adj[u].size(); ++i) {
            int v = adj[u][i];
            if (!vis[v] && dfs(v, d + 1)) return true;
        }
        vis[u] =false;
        return false;
        
    }
    string hasFriendshipChain(int N, vector<int> A, vector<int> B) {
        for (int i = 0; i < N; ++i) adj[i].clear();
        for (int i = 0; i < A.size(); ++i) {
            adj[A[i]].push_back(B[i]);
            adj[B[i]].push_back(A[i]);
        }
        for (int i = 0; i < N; ++i) 
            if (dfs(i, 1)) return "Yay!";
        return ":(";
    }
};

// CUT begin
ifstream data("SmilesTheFriendshipUnicorn.sample");

string next_line() {
    string s;
    getline(data, s);
    return s;
}

template <typename T> void from_stream(T &t) {
    stringstream ss(next_line());
    ss >> t;
}

void from_stream(string &s) {
    s = next_line();
}

template <typename T> void from_stream(vector<T> &ts) {
    int len;
    from_stream(len);
    ts.clear();
    for (int i = 0; i < len; ++i) {
        T t;
        from_stream(t);
        ts.push_back(t);
    }
}

template <typename T>
string to_string(T t) {
    stringstream s;
    s << t;
    return s.str();
}

string to_string(string t) {
    return "\"" + t + "\"";
}

bool do_test(int N, vector<int> A, vector<int> B, string __expected) {
    time_t startClock = clock();
    SmilesTheFriendshipUnicorn *instance = new SmilesTheFriendshipUnicorn();
    string __result = instance->hasFriendshipChain(N, A, B);
    double elapsed = (double)(clock() - startClock) / CLOCKS_PER_SEC;
    delete instance;

    if (__result == __expected) {
        cout << "PASSED!" << " (" << elapsed << " seconds)" << endl;
        return true;
    }
    else {
        cout << "FAILED!" << " (" << elapsed << " seconds)" << endl;
        cout << "           Expected: " << to_string(__expected) << endl;
        cout << "           Received: " << to_string(__result) << endl;
        return false;
    }
}

int run_test(bool mainProcess, const set<int> &case_set, const string command) {
    int cases = 0, passed = 0;
    while (true) {
        if (next_line().find("--") != 0)
            break;
        int N;
        from_stream(N);
        vector<int> A;
        from_stream(A);
        vector<int> B;
        from_stream(B);
        next_line();
        string __answer;
        from_stream(__answer);

        cases++;
        if (case_set.size() > 0 && case_set.find(cases - 1) == case_set.end())
            continue;

        cout << "  Testcase #" << cases - 1 << " ... ";
        if ( do_test(N, A, B, __answer)) {
            passed++;
        }
    }
    if (mainProcess) {
        cout << endl << "Passed : " << passed << "/" << cases << " cases" << endl;
        int T = time(NULL) - 1469718104;
        double PT = T / 60.0, TT = 75.0;
        cout << "Time   : " << T / 60 << " minutes " << T % 60 << " secs" << endl;
        cout << "Score  : " << 300 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT)) << " points" << endl;
    }
    return 0;
}

int main(int argc, char *argv[]) {
    cout.setf(ios::fixed, ios::floatfield);
    cout.precision(2);
    set<int> cases;
    bool mainProcess = true;
    for (int i = 1; i < argc; ++i) {
        if ( string(argv[i]) == "-") {
            mainProcess = false;
        } else {
            cases.insert(atoi(argv[i]));
        }
    }
    if (mainProcess) {
        cout << "SmilesTheFriendshipUnicorn (300 Points)" << endl << endl;
    }
    return run_test(mainProcess, cases, argv[0]);
}
// CUT end

File: SRM 682/SmilesTheFriendshipUnicorn.java
import java.io.*;
import java.util.*;

public class SmilesTheFriendshipUnicorn {
	private List<Integer>[] adj;
	public String hasFriendshipChain(int N, int[] A, int[] B) {
		adj = new ArrayList[N];
		for (int i = 0; i < N; ++i) {
			adj[i] = new ArrayList<Integer>();
		}
		for (int i = 0; i < A.length; ++i) {
			int u = A[i], v = B[i];
			adj[u].add(v);
			adj[v].add(u);
		}
		for (int a = 0; a < N; ++a)
			for (int b : adj[a])
				if (a != b)
					for (int c : adj[b])
						if (c != b && c != a)
							for (int d : adj[c])
								if (d != a && d != b && d != c)
									for (int e : adj[d])
										if(e != a && e != b && e != c && e != d)
											return "Yay!";
		return ":(";
	}

// CUT begin
	public static void main(String[] args){
		System.err.println("SmilesTheFriendshipUnicorn (300 Points)");
		System.err.println();
		HashSet<Integer> cases = new HashSet<Integer>();
        for (int i = 0; i < args.length; ++i) cases.add(Integer.parseInt(args[i]));
        runTest(cases);
	}

	static void runTest(HashSet<Integer> caseSet) {
	    int cases = 0, passed = 0;
	    while (true) {
	    	String label = Reader.nextLine();
	    	if (label == null || !label.startsWith("--"))
	    		break;

            int N = Integer.parseInt(Reader.nextLine());
            int[] A = new int[Integer.parseInt(Reader.nextLine())];
            for (int i = 0; i < A.length; ++i)
                A[i] = Integer.parseInt(Reader.nextLine());
            int[] B = new int[Integer.parseInt(Reader.nextLine())];
            for (int i = 0; i < B.length; ++i)
                B[i] = Integer.parseInt(Reader.nextLine());
            Reader.nextLine();
            String __answer = Reader.nextLine();

            cases++;
            if (caseSet.size() > 0 && !caseSet.contains(cases - 1))
                continue;
    		System.err.print(String.format("  Testcase #%d ... ", cases - 1));

            if (doTest(N, A, B, __answer))
                passed++;
	    }
	    if (caseSet.size() > 0) cases = caseSet.size();
        System.err.println(String.format("%nPassed : %d/%d cases", passed, cases));

        int T = (int)(System.currentTimeMillis() / 1000) - 1469720308;
        double PT = T / 60.0, TT = 75.0;
        System.err.println(String.format("Time   : %d minutes %d secs%nScore  : %.2f points", T / 60, T % 60, 300 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT))));
	}

	static boolean doTest(int N, int[] A, int[] B, String __expected) {
		long startTime = System.currentTimeMillis();
		Throwable exception = null;
		SmilesTheFriendshipUnicorn instance = new SmilesTheFriendshipUnicorn();
		String __result = "";
		try {
			__result = instance.hasFriendshipChain(N, A, B);
		}
		catch (Throwable e) { exception = e; }
		double elapsed = (System.currentTimeMillis() - startTime) / 1000.0;

		if (exception != null) {
			System.err.println("RUNTIME ERROR!");
			exception.printStackTrace();
			return false;
		}
		else if (__expected.equals(__result)) {
			System.err.println("PASSED! " + String.format("(%.2f seconds)", elapsed));
			return true;
		}
		else {
			System.err.println("FAILED! " + String.format("(%.2f seconds)", elapsed));
			System.err.println("           Expected: " + "\"" + __expected + "\"");
			System.err.println("           Received: " + "\"" + __result + "\"");
			return false;
		}
	}

	static class Reader {
        private static final String dataFileName = "/Users/resry/topcoder/SRM 682/SmilesTheFriendshipUnicorn.sample";
	    private static BufferedReader reader;

	    public static String nextLine() {
	        try {
                if (reader == null) {
                    reader = new BufferedReader(new InputStreamReader(new FileInputStream(dataFileName)));
                }
                return reader.readLine();
	        }
	        catch (IOException e) {
	            System.err.println("FATAL!! IOException");
	            e.printStackTrace();
	            System.exit(1);
	        }
	        return "";
	    }
	}
// CUT end
}
